### ISSUE TITLE
FileManager.filePermits semaphore is unfair, blocking scans under high load

### ISSUE THREAD

**Issue by dlmarion** [apache/accumulo#3198]
User reported seeing [messages](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/fs/FileManager.java#L268) in their logs where it was taking hours to reserve files for scan. Prior to 1.9.0 the FileManager.filePermits semaphore was fair, but was changed in [this](https://github.com/apache/accumulo/commit/86d6a69ec8e4e7fe1b187c4a53b8c0a88921ab27) commit. The commit message says, "Also in this change the file managers semaphore was switched to non-fair.  The fair semaphore is expensive and can cause contention." Fair locks and semaphores have a penalty, which seems unnecessary when not running in a high contention heavy-load situation. However, when in that situation, they allow for many threads to make some progress vs a small number of threads making progress and the others doing nothing.

@keith-turner - would like your thoughts on this as you were the author of the referenced commit. Was the penalty for the fair lock severe?

**Comment by ddanielr** [comment]
@dlmarion I can work on this if you want to assign it to me

**Comment by keith-turner** [comment]
> @keith-turner - would like your thoughts on this as you were the author of the referenced commit. Was the penalty for the fair lock severe?

I can not remember the specifics of what I was doing.  I don't know if my comment was based on performance testing or just reading about the performance for the fair semaphore.  Given the bad situation that is happening sometimes, it seems like it would be best to switch it to fair.

Should this change need to be made in 1.10.x?

**Comment by dlmarion** [comment]
> Should this change need to be made in 1.10.x?

Maybe. I'll let @ddanielr evaluate that. If it's easy, then we should probably do that.

**Comment by ddanielr** [comment]
> > Should this change need to be made in 1.10.x?
> 
> Maybe. I'll let @ddanielr evaluate that. If it's easy, then we should probably do that.

I"m still looking into this code so I don't have a complete solution yet. 

The FileManager code is extremely similar between 1.10 and 2.1 so if the fix is constrained to just that class then I don't see why it couldn't be released to 1.10.x considering this problem's impact. 

@dlmarion do you know if the user was experiencing these extended scan blocks while using [Scan Executors](https://accumulo.apache.org/docs/2.x/administration/scan-executors)?

**Comment by dlmarion** [comment]
Yes, it was during scanning

**Comment by ddanielr** [comment]
From the [java docs](https://devdocs.io/openjdk~11/java.base/java/util/concurrent/semaphore): 
```
Generally, semaphores used to control resource access should be initialized as fair, to ensure that no thread is starved out from accessing a resource. When using semaphores for other kinds of synchronization control, the throughput advantages of non-fair ordering often outweigh fairness considerations.
```
Thread starvation for scans seem extremely problematic in a high load environment. Especially so once you add in expected behavior of the Scan Executors thread pools.

A fair semaphore ensures this can't happen by enforcing a FIFO queue for resource requests once all of the available file permits are reserved. 

This code is pretty stable between 1.10 and 2.1.x so given the impact to scans, this bug should be fixed in the 1.10.x line.

**Comment by ctubbsii** [comment]
I added 2.1.1 as a version in which this is fixed, since the fix for this will appear in both the 1.10.3 and the 2.1.1 patch releases.
